### PR TITLE
Audit v2 vault remediation

### PR DESCRIPTION
**Pull request by Krishnakumarskr**
1. [[Centralization Risk in Vault System Leading to Deposit Lock](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/aa41a960-31b1-4e57-b597-8c1cd5e52949)](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/aa41a960-31b1-4e57-b597-8c1cd5e52949)
    
    To be discussed
    
    - For the Custodian deposit - We are using a VASP (e.g Circle) to custody our funds.  We will use the VASPs security and recovery mechanisms.
    - For the withdraw ERC20 - Need recommendations for what can be done here!
        - In the case of ERC20 NOT the vault Asset - Should we change this in such a way that the ADMIN can only withdraw ERC20 tokens that aren’t the Vault Asset.  For example in a USDC vault, the ADMIN can only use this feature for tokens that aren’t USDC.
        - In the case of ERC20 is the vault Asset - someone does a direct transfer (instead of calling Deposit). How those scenarios can be handled?
    
2. [[Risk of Ownership Control Loss in Owner-Dependent Contracts](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/2add75ab-d0d1-4a9f-9dd9-a776c8d2c330)](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/2add75ab-d0d1-4a9f-9dd9-a776c8d2c330)
   
    
    Fixed!
    
3. [[Floating Pragma](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/0884bb2e-ea27-45d3-a4b9-94b5182a7c41)](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/0884bb2e-ea27-45d3-a4b9-94b5182a7c41)
    
    Acknowledged!
    
4. [[Missing Validation for Time Windows in WindowPlugin Contract](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/6ca4b3a8-0523-464e-b2df-63e140f1e42f)](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/6ca4b3a8-0523-464e-b2df-63e140f1e42f)
    
Fixed!
   
    
5. [[Redundant Fallback and Receive Functionality](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/4f904980-5e37-4ba8-835a-da54f5c4ef6d)](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/4f904980-5e37-4ba8-835a-da54f5c4ef6d)
    
    Fixed!

    
6. [[Lack of Configuration Events](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/1032f360-ba4e-4836-9685-70bfb34cebea)](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/1032f360-ba4e-4836-9685-70bfb34cebea)
    
  Fixed!
    
7. [[State Variables That Should Be Immutable](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/03715733-cdef-4b78-ac2b-5f6aecea3721)](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/03715733-cdef-4b78-ac2b-5f6aecea3721)
    
    Fixed!
    
8. [[Missing Zero Addresses Validations Leading to Misconfiguration](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/2390b385-ad80-4930-9fd5-dcaddb171466)](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/2390b385-ad80-4930-9fd5-dcaddb171466)
    
    Fixed!
    
9. [[Missing Validation for collateralPercentage and twap in UpsideVault Contract](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/5c8b1b5f-babb-4362-9fee-fa9bea7a0a2f)](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/5c8b1b5f-babb-4362-9fee-fa9bea7a0a2f)
    
    Fixed for collateral Percentage.
    
    Twap can be any value!
    
    Renamed collateral percentage to upsidePercentage!
    
10. [[Violation of Checks-Effects-Interactions Pattern](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/43b56c59-152c-49ae-a006-72ca1e09a19e)](https://portal.hacken.io/App/Projects/Details/fa0e97cc-7852-4ffa-97d3-94050090d565/Finding/43b56c59-152c-49ae-a006-72ca1e09a19e)
    
    Acknowledged!
    Can someone from the team help us in exactly what need to be changed here?

### CODE REVIEW

File: packages/contracts/src/plugin/WhiteListPlugin.sol [hunk 1]
```diff
@@ -18,27 +18,27 @@ abstract contract WhiteListPlugin {
     }
 
     /// @notice - Address of the White List Provider.
-    IWhiteListProvider public whiteListProvider;
+    IWhiteListProvider public immutable WHITELIST_PROVIDER;
```
Comment (jplodge-pro) [inline packages/contracts/src/plugin/WhiteListPlugin.sol:21]: All uppercase is for [constants](https://docs.soliditylang.org/en/v0.8.26/style-guide.html#constants) not `immutable` state variables (which are just state variables).

**Review by jplodge-pro — CHANGES_REQUESTED** [review]

File: packages/contracts/src/vault/MaturityVault.sol [hunk 2]
```diff
@@ -73,9 +78,8 @@ abstract contract MaturityVault is Vault {
     /**
      * @notice Enables/disables the Maturity Check according to the [status] value.
      * @dev 'Toggling' means flipping the existing state. This is simply a mutator.
-     * @param status Boolean value to toggle
      */
-    function _toggleMaturityCheck(bool status) internal {
-        checkMaturity = status;
```
Comment (lucasia) [inline packages/contracts/src/vault/MaturityVault.sol:79]: similar, let's keep the function taking a bool as an argument and rename it to be setMaturityCheck(bool _checkMaturity)

**Review by lucasia — CHANGES_REQUESTED** [review]
Please kindly update as per suggestions.

File: packages/contracts/src/plugin/MaxCapPlugin.sol [hunk 3]
```diff
@@ -6,6 +6,12 @@ pragma solidity ^0.8.20;
 abstract contract MaxCapPlugin {
     error CredbullVault__MaxCapReached();
```
[the remark this answers is not in the document]
Comment (Krishnakumarskr) [reply packages/contracts/src/plugin/MaxCapPlugin.sol:7]: It's to set the maximum deposit amount that can be deposited by the users collectively to the vault.
ERC20Capped doesn't fit here as it is related to setting maxcap to mint ERC20 tokens.

File: packages/contracts/src/vault/FixedYieldVault.sol [hunk 4]
```diff
@@ -79,23 +93,23 @@ contract FixedYieldVault is MaturityVault, WhiteListPlugin, WindowPlugin, MaxCap
...
+    function toggleWhiteListCheck() public onlyRole(DEFAULT_ADMIN_ROLE) {
+        _toggleWhiteListCheck();
     }
 
     /// @notice Toggle check for window
-    function toggleWindowCheck(bool status) public onlyRole(DEFAULT_ADMIN_ROLE) {
-        _toggleWindowCheck(status);
+    function toggleWindowCheck() public onlyRole(DEFAULT_ADMIN_ROLE) {
+        _toggleWindowCheck();
     }
```
Comment (lucasia) [inline packages/contracts/src/vault/FixedYieldVault.sol:108]: this should follow the same convention - take a bool status and be renamed from toggleWindowCheck to setWindowCheck

**Review by lucasia — APPROVED** [review]

**Pull request merged**